src/webgpu/shader/validation/shader_io/workgroup_size.spec.ts

```diff
@@ -0,0 +1,153 @@
+export const description = `Validation tests for workgroup_size`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { keysOf } from '../../../../common/util/data_tables.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kWorkgroupSizeTests = {
+  x_only_float: {
+    src: `@workgroup_size(8f)`,
+    pass: false,
+  },
+  xy_only_float: {
+    src: `@workgroup_size(8, 8f)`,
+    pass: false,
+  },
+  xyz_float: {
+    src: `@workgroup_size(8, 8, 8f)`,
+    pass: false,
+  },
+  empty: {
+    src: `@workgroup_size()`,
+    pass: false,
+  },
+  empty_x: {
+    src: `@workgroup_size(, 8)`,
+    pass: false,
+  },
+  empty_y: {
+    src: `@workgroup_size(8, , 8)`,
+    pass: false,
+  },
+  invalid_entry: {
+    src: `@workgroup_size(let)`,
+    pass: false,
+  },
+
+  x_only_abstract: {
+    src: `@workgroup_size(8)`,
+    pass: true,
+  },
+  xy_only_abstract: {
+    src: `@workgroup_size(8, 8)`,
+    pass: true,
+  },
+  xyz_abstract: {
+    src: `@workgroup_size(8, 8, 8)`,
+    pass: true,
+  },
+  x_only_unsigned: {
+    src: `@workgroup_size(8u)`,
+    pass: true,
+  },
+  xy_only_unsigned: {
+    src: `@workgroup_size(8u, 8u)`,
+    pass: true,
+  },
+  xyz_unsigned: {
+    src: `@workgroup_size(8u, 8u, 8u)`,
+    pass: true,
+  },
+  x_only_signed: {
+    src: `@workgroup_size(8i)`,
+    pass: true,
+  },
+  xy_only_signed: {
+    src: `@workgroup_size(8i, 8i)`,
+    pass: true,
+  },
+  xyz_signed: {
+    src: `@workgroup_size(8i, 8i, 8i)`,
+    pass: true,
+  },
+
+  override: {
+    src: `@id(42) override block_width = 12u;
+@workgroup_size(block_width)`,
+    pass: true,
+  },
+  override_no_default: {
+    src: `override block_width: i32;
+@workgroup_size(block_width)`,
+    pass: true,
+  },
+
+  trailing_comma_x: {
+    src: `@workgroup_size(8, )`,
+    pass: true,
+  },
+  trailing_comma_y: {
+    src: `@workgroup_size(8, 8,)`,
+    pass: true,
+  },
+  trailing_comma_z: {
+    src: `@workgroup_size(8, 8, 8,)`,
+    pass: true,
+  },
+
+  // Mixed abstract is ok
+  mixed_abstract_signed: {
+    src: `@workgroup_size(8, 8i)`,
+    pass: true,
+  },
+  mixed_abstract_unsigned: {
+    src: `@workgroup_size(8u, 8)`,
+    pass: true,
+  },
+  // Mixed signed and unsigned is not
+  mixed_signed_unsigned: {
+    src: `@workgroup_size(8i, 8i, 8u)`,
+    pass: false,
+  },
+
+  zero_x: {
+    src: `@workgroup_size(0)`,
+    pass: false,
+  },
+  zero_y: {
+    src: `@workgroup_size(8, 0)`,
+    pass: false,
+  },
+  zero_z: {
+    src: `@workgroup_size(8, 8, 0)`,
+    pass: false,
+  },
+  negative_x: {
+    src: `@workgroup_size(-8)`,
+    pass: false,
+  },
+  negative_y: {
+    src: `@workgroup_size(8, -8)`,
+    pass: false,
+  },
+  negative_z: {
+    src: `@workgroup_size(8, 8, -8)`,
+    pass: false,
+  },
+
+  max_values: {
+    src: `@workgroup_size(256, 256, 64)`,
+    pass: true,
+  },
```

Comment:
This is still reasonably likely to fail on many implementations due to the number of invocations limit. So maybe we should have a test that validates beyond these numbers?

Reply:
These numbers are the values listed in the WebGPU spec, so I believe they are the most that needs to be supported. These are all validation tests so they won't be executed on actual devices, but the values should compile I believe.

Reply:
maxComputeInvocationsPerWorkgroup is 256 for reference. So (256,256,64) is too large. Not a big deal since it's only a validation test, but likewise we could ensure larger values are allowed (someone could also ask for an adaptor with higher limits).

Reply:
I wasn't sure if requesting more would be portable, because the adapter can reject the higher limit, no? So, we can only validate that the minimum passes, otherwise it's out of our hands?

Or, since that's validation outside WGSL, we can just treat this as an i32 and pass in whatever and leave it for execution tests to catch big values?

Reply:
I expect https://github.com/gpuweb/gpuweb/pull/4073 to be approved. So there is no upper bound to hit for a shader-creation error. So unless you're checking for pipeline errors the test could (but don't have to) cover larger values. Not really asking for a change here just more as an FYI. The main point of the review was the request for expression (not just declaration or literal) based values.